### PR TITLE
Migrate NuGet-backed managers to the V3 API

### LINKED ISSUE #5331: [ENHANCEMENT] Migrate NuGet V2 APIs to V3 APIs
### Please confirm these before moving forward.

- [x] I have searched for my feature proposal and have not found a work-in-progress/duplicate/resolved/discarded issue.
- [x] This improvement refers to an existing feature. If you want to suggest a new feature, please use [this template](https://github.com/Devolutions/UniGetUI/issues/new?labels=new-feature&projects=&template=feature-request.yml&title=%5BFEATURE+REQUEST%5D+%28Enter+your+description+here%29).
- [x] This improvement is not a bug. If you want to report a bug, please use [this template](https://github.com/Devolutions/UniGetUI/issues/new?labels=bug&projects=&template=bug-issue.yml&title=%5BBUG%5D+%28Enter+your+description+here%29).

### Describe the improvement

Hi there I'm Charles Bissonnette, the Engineering Manager for http://NuGet.Org and I would like to bring this issue to the team.  UniGetUI is still relying on V2 NuGet API and our telemetry shows that UniGetUI ends up consuming a huge portion of DataBase's DTU.   It is important that you migrate UniGetUI's direct NuGet V2/OData integration to NuGet V3 for the generic NuGet-backed managers (.NET Tool, Windows PowerShell, and PowerShell 7), while preserving compatibility for feeds that cannot provide V3 resources.

## Current implementation
The shared implementation is primarily in:
- `src/UniGetUI.PackageEngine.Managers.Generic.NuGet/BaseNuGet.cs`
- `src/UniGetUI.PackageEngine.Managers.Generic.NuGet/Internal/NuGetManifestLoader.cs`
- `src/UniGetUI.PackageEngine.Managers.Generic.NuGet/BaseNuGetDetailsHelper.cs`

It currently uses V2/OData endpoints such as `Search()`, `Packages()`, `FindPackagesById()`, `Packages(Id=...,Version=...)`, and `GetUpdates()`, and parses Atom/OData XML responses.

## Proposed implementation steps

1. **Inventory and define compatibility boundaries**
   - Enumerate every direct V2 URL, OData query, XML response parser, and test fixture in the generic NuGet implementation.
   - Keep Chocolatey's `choco` CLI integration separate; do not convert it solely because Chocolatey sources use `/api/v2/`.
   - Identify which configured sources support a NuGet V3 service index and define the fallback behavior for sources that do not.

2. **Introduce a NuGet V3 resource-index client**
   - Add a focused internal client/abstraction that loads `{source}/index.json`.
   - Parse the service index and select resources by advertised `@type`, including `SearchQueryService`, `RegistrationsBaseUrl`, and `PackageBaseAddress/3.0.0` (or compatible versions).
   - Normalize source URLs and cache the resolved resource index per source for the lifetime of the relevant operation/session.
   - Handle unavailable or malformed indexes with useful logging and a controlled V2 fallback where supported.

3. **Migrate package search**
   - Replace V2 `Search()` and `Packages()` URL construction in `BaseNuGet.FindPackages_UnSafe()` with the V3 SearchQueryService.
   - Send the query using V3 parameters (`q`, `skip`, `take`, `prerelease`, and `semVerLevel` where supported).
   - Deserialize V3 JSON instead of extracting `<entry>` elements with regular expressions.
   - Preserve the existing result limit, pagination, prerelease behavior, source association, duplicate handling, and highest-version selection.
   - Preserve the intent of `UseSubstringSearch`; use exact package-ID/version resources where a search endpoint cannot provide equivalent behavior.

4. **Replace batch update lookup**
   - Remove the dependency on V2 `GetUpdates()`, which has no direct V3 equivalent.
   - For each installed package/source group, obtain available versions and relevant registration metadata using V3 PackageBaseAddress and/or RegistrationsBaseUrl resources.
   - Compare versions locally using the existing version semantics, including prerelease filtering and ignoring versions that are not newer.
   - Preserve package scope, source identity, update dates, logging, and failure reporting.
   - Add bounded concurrency and caching so update checks do not create unbounded request fan-out.

5. **Migrate package versions and package content URLs**
   - Replace `FindPackagesById()` in `BaseNuGetDetailsHelper` with the V3 flat-container version index:
     `{packageBaseAddress}/{lowercase-id}/index.json`.
   - Update `NuGetManifestLoader.GetNuPkgUrl()` to use the V3 flat-container package path:
     `{packageBaseAddress}/{lowercase-id}/{lowercase-version}/{lowercase-id}.{lowercase-version}.nupkg`.
   - Account for normalized lowercase IDs and versions, URL escaping, and feeds with nonstandard resource URL layouts.
   - Add tests for IDs and versions containing characters that require normalization or escaping.

6. **Migrate package metadata and details**
   - Replace the V2 `Packages(Id='...',Version='...')` manifest URL in `NuGetManifestLoader` with V3 registration metadata and/or the package `.nuspec` URL.
   - Adapt `BaseNuGetDetailsHelper` to consume V3 JSON metadata and package `.nuspec` content rather than depending on V2 Atom fields such as `<d:Description>`, `<d:Dependencies>`, and `<d:IconUrl>`.
   - Preserve descriptions, authors/publishers, licenses, release notes, update dates, dependencies, icon URLs, installer URLs, installer sizes, and cached manifest behavior.
   - Prefer structured parsing over regular-expression parsing where practical, while retaining robust handling for incomplete package metadata.

7. **Update manager source definitions**
   - Change the .NET Tool default and known source to `https://api.nuget.org/v3/index.json`.
   - Change the PowerShell Gallery sources to their supported V3 service-index endpoint, confirming the endpoint and behavior for both Windows PowerShell and PowerShell 7.
   - Preserve custom-source support and V2 fallback for feeds that do not expose the required V3 resources.
   - Do not change Chocolatey's source behavior unless a compatible V3 migration is explicitly validated.

8. **Add migration and regression tests**
   - Add HTTP fixture tests for service-index parsing and resource selection.
   - Add tests for V3 search JSON, pagination, prerelease filtering, duplicate versions, and malformed responses.
   - Add tests for flat-container version enumeration and package URL generation.
   - Add tests for registration metadata and `.nuspec` parsing, including dependencies, icons, licenses, and missing fields.
   - Add tests for local update comparison and bounded/cached requests.
   - Retain V2 fallback tests for custom feeds where fallback is implemented.
   - Update existing tests that assert V2 URLs, V2 XML, or V2-specific behavior.

9. **Documentation and rollout**
   - Update code comments and documentation that describe the integration as V2/OData-specific.
   - Document supported V3 resources and the fallback behavior.
   - Run the relevant test projects, formatting checks, and build checks.
   - Review behavior against nuget.org, PowerShell Gallery, a V2-only custom feed, and a custom V3 feed before removing any fallback code.

## Acceptance criteria
- The .NET Tool, Windows PowerShell, and PowerShell 7 managers use NuGet V3 resources for supported feeds.
- Search, version listing, package details, package downloads, icons, dependencies, and update detection continue to work.
- Prerelease filtering, pagination, caching, logging, proxy behavior, and source identity remain functional.
- V2-only custom sources do not regress when fallback is available; unsupported feeds fail clearly when neither API is usable.
- Chocolatey's CLI-based behavior remains unchanged.
- Automated tests cover V3 resource discovery, search, metadata, versions, updates, content URLs, and fallback behavior.

### Describe how this improvement could help users

Moving from NuGet V2/OData to V3 would align UniGetUI with the current NuGet server API model, avoid dependence on legacy XML/OData endpoints, improve interoperability with modern NuGet-compatible feeds, and provide a more maintainable foundation for search, metadata, package downloads, and update detection. It should also make feed capabilities explicit through service-index discovery and allow UniGetUI to retain compatibility with older custom repositories where possible.

### PR DESCRIPTION
**Pull request by GabrielDuf**
## Summary                                                                                                                                    
                                                                                                                                                   
Engineering Manager for NuGet.org reported that UniGetUI's V2/OData traffic consumes a large share of nuget.org's database DTU. This moves the NuGet-backed managers onto the V3 API for feeds that support it, while leaving V2 fully intact for feeds that don't.

## What is *not* migrated, and why                                                                                                            
                                                                                                                                                   
**The PowerShell Gallery has no usable V3 service index.** Every path under `powershellgallery.com/api/v3/` returns `403 — blocked by a Web Application Firewall rule` (while `/api/v2/` returns 200, and a nonexistent `/v3/` returns 404, so the prefix exists and is deliberately blocked). Microsoft's own PSResourceGet docs still register PSGallery as `.../api/v2`. Step 7 of the issue isn't actionable for it, so both PowerShell managers stay on V2.                                            
                                                                                                                                           
**Chocolatey** stays on V2 too — `community.chocolatey.org/api/v3/index.json` is a 404. Worth flagging one correction to the issue's framing: Chocolatey is *not* purely CLI-driven. It derives from `BaseNuGet` and `FindPackages_UnSafe` is `sealed`, so its **search, details and icons** run through the shared HTTP path; only updates and version listing are `choco`. V2 is therefore a permanently maintained path, not a fallback to delete later.                                                                                  
                                                                                                                                           
A V3-capable *custom* PowerShell repository (Azure Artifacts, GitHub Packages, JFrog, MyGet) is picked up automatically with no per-manager work.

 ## Risk                                                                                                                                       
                                                                                                                                                   
Highest-risk area is `CompareVersions`, because it touches every manager's update decision, not just NuGet. Mitigations: the default preserves current semantics exactly, and there are explicit tests asserting that revision ecosystems still read a trailing suffix as **newer** — so a future "simplification" that makes the shared parser SemVer-aware will fail CI.                                                                      
                                                                                                                                             
Second: users sitting on a pre-release across eight managers will now be offered the stable release they were previously denied. That's the fix working, but it will look like a surge of new updates.                                            
                                                                                                                                                   
## Behaviour changes users will notice (intended)                                                                                             
                                                                                                                                                   
- `.NET Tool` search returns **fewer** results — libraries that can't be installed as tools are filtered out                                  
- Version dropdowns are **newest-first**, not alphabetical (and no longer truncated: `dotnet-ef` went from 100 to all 266 versions)           
- Details → Manifest URL is now an `api.nuget.org/v3/registration5-…json` URL                                                                 
- `.NET Tool` installs now default to **global**                                                                                              
- A package on a pre-release may suddenly show an update

### CODE REVIEW

File: src/UniGetUI.PackageEngine.Managers.Generic.NuGet/BaseNuGetDetailsHelper.cs [hunk 1]
```diff
@@ -223,8 +230,148 @@ Match match in Regex.Matches(
...
+
+            return true;
+        }
+
+        private static V3CatalogEntry? GetOrFetchCatalogEntry(
+            IPackage package,
+            NuGetV3ServiceIndex index
+        )
+        {
+            long hash = package.GetHash();
```
[the remark this answers is not in the document]
Comment (GabrielDuf) [reply src/UniGetUI.PackageEngine.Managers.Generic.NuGet/BaseNuGetDetailsHelper.cs]: [d6b5b33](https://github.com/Devolutions/UniGetUI/pull/5340/commits/d6b5b33800d1de095a3fc46c01bf1b7ef511133b)

File: src/UniGetUI.PackageEngine.Managers.Generic.NuGet/Internal/NuGetV3Client.cs [hunk 2]
```diff
@@ -0,0 +1,713 @@
...
+
+            newer.Sort((left, right) => right.Parsed.CompareTo(left.Parsed));
+
+            int probes = 0;
+            foreach ((SemanticVersion _, string raw) in newer)
+            {
+                if (probes >= MaxUnlistedProbes)
+                    break;
+
+                probes++;
```
[the remark this answers is not in the document]
Comment (GabrielDuf) [reply src/UniGetUI.PackageEngine.Managers.Generic.NuGet/Internal/NuGetV3Client.cs]: [d6b5b33](https://github.com/Devolutions/UniGetUI/pull/5340/commits/d6b5b33800d1de095a3fc46c01bf1b7ef511133b)

File: src/UniGetUI.PackageEngine.Managers.Generic.NuGet/Internal/NuGetV3Client.cs [hunk 3]
```diff
@@ -0,0 +1,713 @@
...
+                );
+            }
+
+            return null;
+        }
+
+        public static bool IsListed(NuGetV3ServiceIndex index, string packageId, string version)
+        {
+            V3RegistrationLeaf? leaf = GetRegistrationLeaf(index, packageId, version);
+            return leaf?.Listed ?? true;
```
[the remark this answers is not in the document]
Comment (GabrielDuf) [reply src/UniGetUI.PackageEngine.Managers.Generic.NuGet/Internal/NuGetV3Client.cs]: [d6b5b33](https://github.com/Devolutions/UniGetUI/pull/5340/commits/d6b5b33800d1de095a3fc46c01bf1b7ef511133b)

File: src/UniGetUI.PackageEngine.Managers.Generic.NuGet/Internal/NuGetV3Client.cs [hunk 4]
```diff
@@ -0,0 +1,713 @@
...
+                !Uri.TryCreate(
+                    $"{index.PackageBaseAddress.AbsoluteUri.TrimEnd('/')}/{EscapeId(packageId)}/index.json",
+                    UriKind.Absolute,
+                    out Uri? versionsUrl
+                )
+            )
+                return [];
+
+            if (!TryDownloadString(versionsUrl, out string content))
+                return [];
```
[the remark this answers is not in the document]
Comment (GabrielDuf) [reply src/UniGetUI.PackageEngine.Managers.Generic.NuGet/Internal/NuGetV3Client.cs:294]: [d6b5b33](https://github.com/Devolutions/UniGetUI/pull/5340/commits/d6b5b33800d1de095a3fc46c01bf1b7ef511133b)

File: src/UniGetUI.Core.Tools/SemanticVersion.cs [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+            );
+
+            if (leftNumeric && rightNumeric)
+                return leftValue.CompareTo(rightValue);
+            if (leftNumeric)
+                return -1;
+            if (rightNumeric)
+                return 1;
+
+            return string.Compare(left, right, StringComparison.OrdinalIgnoreCase);
```
[the remark this answers is not in the document]
Comment (GabrielDuf) [reply src/UniGetUI.Core.Tools/SemanticVersion.cs]: [d6b5b33](https://github.com/Devolutions/UniGetUI/pull/5340/commits/d6b5b33800d1de095a3fc46c01bf1b7ef511133b)

File: src/UniGetUI.PackageEngine.Managers.Generic.NuGet/Internal/NuGetManifestLoader.cs [hunk 6]
```diff
@@ -27,6 +27,17 @@ public static Uri GetManifestUrl(IPackage package)
         /// <returns>A Uri object</returns>
         public static Uri GetNuPkgUrl(IPackage package)
         {
+            if (NuGetV3ServiceIndex.Resolve(package.Source) is { } index)
```
[the remark this answers is not in the document]
Comment (GabrielDuf) [reply src/UniGetUI.PackageEngine.Managers.Generic.NuGet/Internal/NuGetManifestLoader.cs]: [d6b5b33](https://github.com/Devolutions/UniGetUI/pull/5340/commits/d6b5b33800d1de095a3fc46c01bf1b7ef511133b)

File: src/UniGetUI.PackageEngine.Managers.Generic.NuGet/BaseNuGet.cs [hunk 7]
```diff
@@ -203,6 +236,61 @@ protected sealed override IReadOnlyList<Package> FindPackages_UnSafe(string quer
...
+                    CoreTools.FormatAsName(result.Id),
+                    result.Id,
+                    result.Version,
+                    source,
+                    this
+                );
+                packages.Add(nativePackage);
+
+                if (!result.IsExactIdFallback)
+                    V3IconUrls[nativePackage.GetHash()] = result.IconUrl ?? string.Empty;
```
[the remark this answers is not in the document]
Comment (GabrielDuf) [reply src/UniGetUI.PackageEngine.Managers.Generic.NuGet/BaseNuGet.cs]: premise false for nuget.org; partially honoured at zero cost

File: src/UniGetUI.PackageEngine.Managers.Generic.NuGet/BaseNuGetDetailsHelper.cs [hunk 8]
```diff
@@ -252,13 +399,75 @@ Match match in Regex.Matches(
...
+            if (Uri.TryCreate(entry.IconUrl, UriKind.Absolute, out Uri? iconUrl))
+                return new CacheableIcon(iconUrl, package.VersionString);
+
+            if (
+                !string.IsNullOrWhiteSpace(entry.IconFile)
+                && index.PackageBaseAddress is not null
+                && Uri.TryCreate(
+                    $"{index.PackageBaseAddress.AbsoluteUri.TrimEnd('/')}"
+                        + $"/{NuGetV3Client.EscapeId(package.Id)}"
+                        + $"/{NuGetV3Client.EscapeVersion(package.VersionString)}/icon",
```
[the remark this answers is not in the document]
Comment (GabrielDuf) [reply src/UniGetUI.PackageEngine.Managers.Generic.NuGet/BaseNuGetDetailsHelper.cs]: [7c0b2f0](https://github.com/Devolutions/UniGetUI/pull/5340/commits/7c0b2f0a563d9d609a148f954db52a0ccd9b0704)

File: src/UniGetUI.PackageEngine.Managers.Generic.NuGet/Internal/NuGetV3Client.cs [hunk 9]
```diff
@@ -0,0 +1,780 @@
...
+                Logger.Warn(e);
+                return null;
+            }
+        }
+
+        private static bool IsHigherVersion(string candidate, string current)
+        {
+            if (!SemanticVersion.TryParse(candidate, out SemanticVersion parsedCandidate))
+                return false;
+            if (!SemanticVersion.TryParse(current, out SemanticVersion parsedCurrent))
```
[the remark this answers is not in the document]
Comment (GabrielDuf) [reply src/UniGetUI.PackageEngine.Managers.Generic.NuGet/Internal/NuGetV3Client.cs]: [7c0b2f0](https://github.com/Devolutions/UniGetUI/pull/5340/commits/7c0b2f0a563d9d609a148f954db52a0ccd9b0704)

File: src/UniGetUI.PackageEngine.Managers.Generic.NuGet/Internal/NuGetV3Client.cs [hunk 10]
```diff
@@ -0,0 +1,806 @@
...
+            string query,
+            bool includePreRelease,
+            string? requiredPackageType
+        )
+        {
+            string candidate = query.Trim();
+            if (candidate.Length is 0)
+                return [];
+
+            string? version = SelectHighestVersion(GetVersions(index, candidate), includePreRelease);
```
[the remark this answers is not in the document]
Comment (GabrielDuf) [reply src/UniGetUI.PackageEngine.Managers.Generic.NuGet/Internal/NuGetV3Client.cs]: [5974911](https://github.com/Devolutions/UniGetUI/pull/5340/commits/5974911c5c0a67884bfe0b7124e0a7c30ddee20d)

File: src/UniGetUI.PackageEngine.Managers.Generic.NuGet/Internal/NuGetV3Client.cs [hunk 11]
```diff
@@ -0,0 +1,866 @@
...
+        public static IReadOnlyList<string> GetVersions(
+            NuGetV3ServiceIndex index,
+            string packageId,
+            out bool requestFailed
+        )
+        {
+            requestFailed = false;
+
+            if (index.PackageBaseAddress is null)
+                return [];
```
[the remark this answers is not in the document]
Comment (GabrielDuf) [reply src/UniGetUI.PackageEngine.Managers.Generic.NuGet/Internal/NuGetV3Client.cs:272]: [8fb0ed9](https://github.com/Devolutions/UniGetUI/pull/5340/commits/8fb0ed9bcbdc468aa148e2df814444d2c9bee19d)

File: src/UniGetUI.PackageEngine.Managers.Generic.NuGet/Internal/NuGetV3Client.cs [hunk 12]
```diff
@@ -0,0 +1,873 @@
...
+
+            V3RegistrationLeaf? leaf = GetRegistrationLeaf(index, packageId, version);
+            if (leaf is null)
+                return ListedStatus.Unknown;
+
+            return leaf.Listed switch
+            {
+                false => ListedStatus.Unlisted,
+                _ => ListedStatus.Listed,
+            };
```
[the remark this answers is not in the document]
Comment (GabrielDuf) [reply src/UniGetUI.PackageEngine.Managers.Generic.NuGet/Internal/NuGetV3Client.cs:503]: I sampled 22 leaves across two packages on `registration5-gz-semver2`:                                                                                                                                                       
                                                                                                                                                                                                                               
  The root flag was present in **22/22** sampled leaves and agreed with `catalogEntry.listed` in every case, including the genuinely unlisted versions. So `leaf.Listed` does not remain null on real nuget.org responses, and unlisted versions are correctly rejected — `GetUpdateCandidateSkipsUnlistedVersions` covers that path end to end.
